### ISSUE TITLE
[BUG] (Firefox only) contentEditable text inputs duplicate line breaks

### ISSUE THREAD

**Issue by khanonnie** [SillyTavern/SillyTavern#2777]
### Environment

🪟 Windows

### System

Firefox 130.0b9 (64-bit), Windows 11 LTSC

### Version

staging (58e6ae2fc58f1416b1d6c30a33321a3447f41e25)

### Desktop Information

- Branch: staging
- Backend is Node v18.16.0 on Ubuntu Server

### Describe the problem

After rebasing onto latest staging, I'm seeing an issue with the Advanced Formatting tab where the Context prompt editor and Instruct prompt editor are doing weird things with line breaks.  It appears that a recent transition from `<textarea>` elements to contentEditable `<div>`s is the culprit.

Firefox unfortunately has pretty janky contentEditable behavior when it comes to line breaks. It will insert invisible line breaks when editing text, and when serializing the contents of a contenteditable using `innerText`, it seems to double or triple up every <br> tag in the <div>, so you get two or three `\n` characters in the serialized text per visual line break.

This means that when ST goes to update the config object, it seems to do so via innerText and then all of those extra newlines get saved to the instruct/context prompts and eventually get rendered into the story string sent to the API, making models unhappy.

Also, if you use the "Expand Editor" button to pull open the big popup to edit instruct/context prompts, each time you repeat that process, you will re-duplicate all of the line breaks, eventually leading to dozens or hundreds of them if you keep opening and closing the popup. Best as I can tell that is caused by:

```javascript
// chats.js 
    $(document).on('click', '.editor_maximize', function () {
        const broId = $(this).attr('data-for');
        const bro = $(`#${broId}`);
        const contentEditable = bro.is('[contenteditable]');
        const withTab = $(this).attr('data-tab');

        if (!bro.length) {
            console.error('Could not find editor with id', broId);
            return;
        }

        const wrapper = document.createElement('div');
        wrapper.classList.add('height100p', 'wide100p', 'flex-container');
        wrapper.classList.add('flexFlowColumn', 'justifyCenter', 'alignitemscenter');
        const textarea = document.createElement('textarea');
        textarea.value = String(contentEditable ? bro[0].innerText : bro.val());   // ⬅️ innerText here converts <br> into \n\n on firefox
        textarea.classList.add('height100p', 'wide100p', 'maximized_textarea');
        bro.hasClass('monospace') && textarea.classList.add('monospace');
        textarea.addEventListener('input', function () {
            if (contentEditable) {
                bro[0].innerText = textarea.value;
                bro.trigger('input');
            } else {
                bro.val(textarea.value).trigger('input');
            }
        });
        wrapper.appendChild(textarea);
```
[firefox_2024-09-04_21-38-07.webm](https://github.com/user-attachments/assets/1959a42b-234b-4543-82d0-954e5eb3e0b8)

## Fixes
Still looking into it, ~~but you probably have to use `textContent` instead -- however this may need to be combined with some css properties to preserve whitespace (`white-space: pre-wrap`?)~~
- cannot get `textContent` to return newlines with any combination of whitespace rules
other options I can find include
- setting onKeyDown listeners to intercept Enter key and stop the browser from inserting markup
- use innerHTML instead of innerText and manually render it yourself, maybe showdown can do this? -- probably not that hard since all of ST's contentEditables are plain text and should only have `br`/`p`
  - follow-up, showdown cannot do this, it is also broken (https://github.com/showdownjs/showdown/issues/974)
- idk, this bug sucks

## See Also
https://stackoverflow.com/questions/40417527/how-do-i-preserve-line-breaks-when-getting-text-from-a-textarea/40426477#40426477
https://bugzilla.mozilla.org/show_bug.cgi?id=1615852
#2779 

### Please tick the boxes

- [X] I have explained the issue clearly, and I included all relevant info
- [X] I have checked that this [issue hasn't already been raised](https://github.com/SillyTavern/SillyTavern/issues?q=is%3Aissue)
- [X] I have checked the [docs](https://docs.sillytavern.app/) ![important](https://img.shields.io/badge/Important!-F6094E)

**Comment by RossAscends** [comment]
in testing with textContent I find it does not save new lines whatsoever

**Comment by khanonnie** [comment]
Yeah I'm seeing that as well, the SO answers I've found are not accurate (a lot of them are ancient, probably why).  Trying a few other combinations of things without much luck,
The other fix I'm seeing in a lot of places is the one you mentioned in discord of hijacking the Enter keydown to try to stop the browser from inserting its markup, which is pretty gross.

**Comment by khanonnie** [comment]
@RossAscends I did some dumb thing like this
```javascript
// utils.js
/**
 * Retrieves the content of a textarea or contenteditable element, accounting
 * for browser quirks around line breaks and whitespace. Plain text only
 * @param {HTMLElement} element The textarea or contenteditable element
 * @returns {string} The content of the element
 */
export function getTextInputContent(element) {

    // Don't try to use innerText on contentEditables, there are horrible bugs in Firefox
    const isTextarea = element instanceof HTMLTextAreaElement;
    if (isTextarea) {
        return element.value;
    } else if (element.isContentEditable) {
        const clean = DOMPurify.sanitize(element.innerHTML, { ALLOWED_TAGS: ['br', 'p', 'div'] });
        const div = document.createElement('div');
        div.innerHTML = clean;

        function renderPlainText(node) {
            let result = '';
            if (node.nodeType === Node.TEXT_NODE) {
                result += node.textContent;
            } else if (node.nodeType === Node.ELEMENT_NODE) {
                if (node.nodeName === 'BR') {
                    result += '\n';
                } else if (['P', 'DIV'].includes(node.nodeName)) {
                    result += Array.from(node.childNodes).map(renderPlainText).join('') + '\n';
                }
            }
            return result;
        }

        return Array.from(div.childNodes).map(renderPlainText).join('');
    }
}
```
And it seems to handle pasting between textareas, dealing with firefox's weird behavior of creating extra `<div><br></div>`, block nodes (I noticed Edge seemed to do this in some cases too), different behavior between browsers when initializing the contenteditable's content, etc.

Problem is you have to call it any time you want to get content out of one of these contenteditables, but that seems better to setting event listeners on keydown since you only have to do this work once when getting content out of these, which afaik is always debounced.  Still feels janky as hell but I could not find any better solutions -- with all alternative approaches I tried (such as pulling the HTML out of the contenteditable and putting it into a new dummy node and innerText'ing that instead) Firefox does weird stuff with the line breaks.

Edit: nvm there are still some other odd issues with this, it needs to only handle line breaks at depth 0 because otherwise firefox inserts `<div><br></div>` on a single line break input which is converted into `\n\n`.

**Comment by khanonnie** [comment]
Maybe fixed in #2779?  It seems to work okay for me on firefox/edge on staging branch but I have not tested extensively. I think @RossAscends has some WIP stuff going on around these contenteditable text areas, so if it fixes the problem feel free to just steal the commit in that PR and close it.

**Comment by Cohee1207** [comment]
Resolution: reverted back to textareas. Chromium auto-height lags are compensated by other means.
